### PR TITLE
RemovePodsViolatingTopologySpreadConstraint Plugin

### PR DESCRIPTION
**Pull request by jklaw90**

### CODE REVIEW

**Comment by JaneLiuL** [comment]
thanks for your pr,  when you finish you can ping me :)

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go [hunk 1]
```diff
@@ -47,27 +48,65 @@ type topology struct {
...
+// RemovePodsViolatingTopologySpreadConstraint evicts pods which violate their topology spread constraints
+type RemovePodsViolatingTopologySpreadConstraint struct {
+	handle         framework.Handle
+	args           *componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs
+	taintFilterFnc func(taint *v1.Taint) bool
+	evictorFilter  *evictions.EvictorFilter
+	podFilter      podutil.FilterFunc
+}
+
+var _ framework.DeschedulePlugin = &RemovePodsViolatingTopologySpreadConstraint{}
```
Comment (a7i) [inline pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go:60]: I believe this was intended to be a `BalancePlugin` as opposed to `DeschedulePlugin`

See a reference to the [proposal](https://github.com/kubernetes-sigs/descheduler/issues/837#issuecomment-1148878383)

Deschedule Plugin does not takes other pods into account, which we know that topology spread constraint selector does.

File: test/e2e/e2e_topologyspreadconstraint_test.go [hunk 2]
```diff
@@ -82,27 +84,31 @@ func TestTopologySpreadConstraint(t *testing.T) {
...
-					false,
-				),
 				getPodsAssignedToNode,
+				true,
+				false,
+				false,
+				false,
+			)
+
+			plugin, err := removepodsviolatingnodetaints.New(&componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs{
```
Comment (a7i) [inline test/e2e/e2e_topologyspreadconstraint_test.go:97]:
```suggestion
			plugin, err := New(&componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs{
```

**Comment by jklaw90** [comment]
@JaneLiuL

File: pkg/apis/componentconfig/types_pluginargs.go [hunk 3]
```diff
@@ -78,3 +78,17 @@ type RemoveDuplicatesArgs struct {
...
+
+// RemovePodsViolatingTopologySpreadConstraintArgs holds arguments used to configure RemovePodsViolatingTopologySpreadConstraint plugin.
+type RemovePodsViolatingTopologySpreadConstraintArgs struct {
+	metav1.TypeMeta
+
+	Namespaces                 *api.Namespaces
+	LabelSelector              *metav1.LabelSelector
+	IncludeSoftConstraints     bool
+	ThresholdPriority          *int32
+	ThresholdPriorityClassName string
```
Comment (a7i) [inline pkg/apis/componentconfig/types_pluginargs.go:92]:
```suggestion
````
This is now done at a higher level.
Validation is now checked [here](https://github.com/kubernetes-sigs/descheduler/blob/83c4f5d5261e7113b6367c70914cedd1af7792ff/pkg/descheduler/descheduler.go#L371-L379)

And it's added to EvictionFilter [here](https://github.com/kubernetes-sigs/descheduler/blob/83c4f5d5261e7113b6367c70914cedd1af7792ff/pkg/descheduler/descheduler.go#L389)

File: pkg/descheduler/strategy_migration.go [hunk 4]
```diff
@@ -151,4 +152,27 @@ var pluginsMap = map[string]func(ctx context.Context, nodes []*v1.Node, params *
 			klog.V(1).ErrorS(err, "plugin finished with error", "pluginName", removeduplicates.PluginName)
 		}
 	},
+	"RemovePodsViolatingTopologySpreadConstraint": func(ctx context.Context, nodes []*v1.Node, params *api.StrategyParameters, handle *handleImpl) {
+		args := &componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs{
+			Namespaces:                 params.Namespaces,
+			LabelSelector:              params.LabelSelector,
+			IncludeSoftConstraints:     params.IncludePreferNoSchedule,
+			ThresholdPriority:          params.ThresholdPriority,
+			ThresholdPriorityClassName: params.ThresholdPriorityClassName,
```
Comment (a7i) [inline pkg/descheduler/strategy_migration.go:161]:
```suggestion
````

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go [hunk 5]
```diff
@@ -371,8 +374,10 @@ func TestTopologySpreadConstraint(t *testing.T) {
 				},
 			}),
 			expectedEvictedCount: 66,
-			strategy:             api.DeschedulerStrategy{},
-			namespaces:           []string{"ns1"},
+			strategy: api.DeschedulerStrategy{
```
Comment (a7i) [inline pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go:377]: wouldn't it be better to just pass in `componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs` as opposed to doing the mapping in the unit test?

File: pkg/apis/componentconfig/types_pluginargs.go [hunk 6]
```diff
@@ -78,3 +78,17 @@ type RemoveDuplicatesArgs struct {
...
+	NodeFit                    bool
```
Comment (a7i) [inline pkg/apis/componentconfig/types_pluginargs.go:93]:
```suggestion
```

I'm pretty sure `NodeFit` is also no longer needed. Added to [EvictionFilter](https://github.com/kubernetes-sigs/descheduler/blob/83c4f5d5261e7113b6367c70914cedd1af7792ff/pkg/descheduler/descheduler.go#L388) here

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go (hunk 5 above)
Comment (jklaw90) [reply pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go]: good call, i forgot to come back to this.

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go [hunk 7]
```diff
@@ -47,30 +48,44 @@ type topology struct {
...
-) {
-	strategyParams, err := validation.ValidateAndParseStrategyParams(ctx, client, strategy.Params)
+// RemovePodsViolatingTopologySpreadConstraint evicts pods which violate their topology spread constraints
+type RemovePodsViolatingTopologySpreadConstraint struct {
+	handle    framework.Handle
+	args      *componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs
+	podFilter podutil.FilterFunc
+}
+
+var _ framework.BalancePlugin = &RemovePodsViolatingTopologySpreadConstraint{}
```
Comment (ingvagabund) [inline pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go:58]: Missing `var _ framework.Plugin = &RemovePodsViolatingTopologySpreadConstraint{}` check

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go [hunk 8]
```diff
@@ -63,12 +67,9 @@ func TestTopologySpreadConstraint(t *testing.T) {
...
 			expectedEvictedCount: 0,
-			strategy: api.DeschedulerStrategy{
-				Params: &api.StrategyParameters{
-					NodeFit: false,
-				},
-			},
-			namespaces: []string{"ns1"},
+			namespaces:           []string{"ns1"},
+			args:                 componentconfig.RemovePodsViolatingTopologySpreadConstraintArgs{},
+			nodeFit:              false,
```
Comment (ingvagabund) [inline pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint_test.go:72]: This line can be omitted as `false` is the default value for `nodeFit`. The same for other occurrences.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

File: pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go (hunk 7 above)
Comment (jklaw90) [reply pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go:58]: doesn't BalancePlugin implement it? is it necessary?

Comment (ingvagabund) [reply pkg/framework/plugins/removepodsviolatingtopologyspreadconstraint/topologyspreadconstraint.go:58]: You are right :) Please ignore my comment.

**Comment by ingvagabund** [comment]
@jklaw90 would you please rebase and squash the commits before the final review?

**Comment by jklaw90** [comment]
@ingvagabund should be good to go now, thanks!

**Pull request merged**